{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Install and Import Necessary Tools#\n",
    "\n",
    "pip install selenium\n",
    "pip install chromedriver\n",
    "pip install webdriver_manager "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "from webdriver_manager.chrome import ChromeDriverManager\n",
    "from selenium.webdriver.support.ui import Select\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from webdriver_manager.chrome import ChromeDriverManager\n",
    "from selenium.common.exceptions import NoSuchElementException"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         Gene ID              sequence           \n",
      "guide                                            \n",
      "0           SHAM  CTTTCCTTCCCAGGGCGTGA       sham\n",
      "1           ACTB  CTTTCCTTCCCAGGGCGTGA  duplicate\n",
      "2           ACTB  CTTTCCTTCCCAGGGCGTGA        NaN\n",
      "3      HNRNPA2B1  AGCCACCAATAAAGAGCTTA        NaN\n",
      "4           SND1  CCCCTACGTAGGTCCTCTCA        NaN\n"
     ]
    }
   ],
   "source": [
    "##Read in CSV containing Guides##\n",
    "\n",
    "guides = pd.read_csv('Sample_Input.csv', \n",
    "            index_col='guide', \n",
    "            header=0, \n",
    "            names=['guide', 'Gene ID', 'sequence', ''])\n",
    "\n",
    "print(guides) #the first row is a \"SHAM\" row to allow the genome browser to load before the girst guide is run\n",
    "                #SHAM row is later removed before being fed into second function\n",
    "                #duplicate row is to account for lack of out for whatever the first input is"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## WORKING FULL 1 - With Loop ##\n",
    "\n",
    "def get_genomic_location():\n",
    "    \n",
    "    from selenium import webdriver\n",
    "    from selenium.webdriver.common.keys import Keys\n",
    "    from webdriver_manager.chrome import ChromeDriverManager\n",
    "    from selenium.common.exceptions import NoSuchElementException\n",
    "    import time\n",
    "\n",
    "    driver = webdriver.Chrome(ChromeDriverManager().install()) #define driver, I use Chrome, then load url\n",
    "    driver.get(\"https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr7%3A5528404%2D5530403&hgsid=815674809_ypAZM1pVZHGgvsosmAcenVUFkPzy\")\n",
    "    assert \"Human hg38\" in driver.title\n",
    "    driver.minimize_window()\n",
    "\n",
    "    chrsm_loc = [] #empty list where genomic locations will go into\n",
    "    \n",
    "    for i in guides[\"sequence\"]: #feed in the guide from the dataframe through the loop\n",
    "        \n",
    "        time.sleep(10)\n",
    "        elem = driver.find_element_by_name(\"hgt.positionInput\") #find search bar\n",
    "        elem.clear() #clear search bar\n",
    "        elem.send_keys(i) #enter guide\n",
    "        elem.send_keys(Keys.RETURN) #enter aka click\n",
    "        try: \n",
    "            browser = driver.find_element_by_partial_link_text('browser') #return to genome browser\n",
    "            browser.send_keys(Keys.RETURN)\n",
    "            \n",
    "            zoom = driver.find_element_by_name(\"hgt.out4\") #zoom out by 100x for 2,000bp\n",
    "            zoom.send_keys(Keys.RETURN)\n",
    "            \n",
    "            time.sleep(12)\n",
    "            \n",
    "            chrsm = driver.find_element_by_id(\"positionDisplay\").text #take genomic position as text\n",
    "            #print(chrsm)\n",
    "            chrsm_loc.append(chrsm) #add it to the empty list\n",
    "            loc = pd.DataFrame(chrsm_loc) #make it into a data frame in order to add to initial csv -> dataframe\n",
    "            guides[\"Genomic_Location\"] = loc\n",
    "            print(guides)\n",
    "\n",
    "        except NoSuchElementException:  #this is for guides that do not return a match from UCSC genome browser\n",
    "            chrsm = \"NA\" #enter NA for failed ones and append to that guide in the dataframe\n",
    "            chrsm_loc.append(chrsm)\n",
    "            loc = pd.DataFrame(chrsm_loc)\n",
    "            guides[\"Genomic_Location\"] = loc\n",
    "            print(guides)\n",
    "            pass\n",
    "            driver.back() #go back one page to start over\n",
    "            continue\n",
    "        \n",
    "    driver.close()\n",
    "    \n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Looking for [chromedriver 80.0.3987.106 mac64] driver in cache \n",
      "File found in cache by path [/Users/santiagoacero/.wdm/drivers/chromedriver/80.0.3987.106/mac64/chromedriver]\n",
      "         Gene ID              sequence            Genomic_Location\n",
      "guide                                                             \n",
      "0           SHAM  CTTTCCTTCCCAGGGCGTGA       sham               NA\n",
      "1           ACTB  CTTTCCTTCCCAGGGCGTGA  duplicate              NaN\n",
      "2           ACTB  CTTTCCTTCCCAGGGCGTGA        NaN              NaN\n",
      "3      HNRNPA2B1  AGCCACCAATAAAGAGCTTA        NaN              NaN\n",
      "4           SND1  CCCCTACGTAGGTCCTCTCA        NaN              NaN\n",
      "         Gene ID              sequence                     Genomic_Location\n",
      "guide                                                                      \n",
      "0           SHAM  CTTTCCTTCCCAGGGCGTGA       sham                        NA\n",
      "1           ACTB  CTTTCCTTCCCAGGGCGTGA  duplicate  chr7:5,528,404-5,530,403\n",
      "2           ACTB  CTTTCCTTCCCAGGGCGTGA        NaN                       NaN\n",
      "3      HNRNPA2B1  AGCCACCAATAAAGAGCTTA        NaN                       NaN\n",
      "4           SND1  CCCCTACGTAGGTCCTCTCA        NaN                       NaN\n",
      "         Gene ID              sequence                     Genomic_Location\n",
      "guide                                                                      \n",
      "0           SHAM  CTTTCCTTCCCAGGGCGTGA       sham                        NA\n",
      "1           ACTB  CTTTCCTTCCCAGGGCGTGA  duplicate  chr7:5,528,404-5,530,403\n",
      "2           ACTB  CTTTCCTTCCCAGGGCGTGA        NaN  chr7:5,528,404-5,530,403\n",
      "3      HNRNPA2B1  AGCCACCAATAAAGAGCTTA        NaN                       NaN\n",
      "4           SND1  CCCCTACGTAGGTCCTCTCA        NaN                       NaN\n",
      "         Gene ID              sequence                       Genomic_Location\n",
      "guide                                                                        \n",
      "0           SHAM  CTTTCCTTCCCAGGGCGTGA       sham                          NA\n",
      "1           ACTB  CTTTCCTTCCCAGGGCGTGA  duplicate    chr7:5,528,404-5,530,403\n",
      "2           ACTB  CTTTCCTTCCCAGGGCGTGA        NaN    chr7:5,528,404-5,530,403\n",
      "3      HNRNPA2B1  AGCCACCAATAAAGAGCTTA        NaN  chr7:26,196,703-26,198,702\n",
      "4           SND1  CCCCTACGTAGGTCCTCTCA        NaN                         NaN\n",
      "         Gene ID              sequence             \\\n",
      "guide                                               \n",
      "0           SHAM  CTTTCCTTCCCAGGGCGTGA       sham   \n",
      "1           ACTB  CTTTCCTTCCCAGGGCGTGA  duplicate   \n",
      "2           ACTB  CTTTCCTTCCCAGGGCGTGA        NaN   \n",
      "3      HNRNPA2B1  AGCCACCAATAAAGAGCTTA        NaN   \n",
      "4           SND1  CCCCTACGTAGGTCCTCTCA        NaN   \n",
      "\n",
      "                   Genomic_Location  \n",
      "guide                                \n",
      "0                                NA  \n",
      "1          chr7:5,528,404-5,530,403  \n",
      "2          chr7:5,528,404-5,530,403  \n",
      "3        chr7:26,196,703-26,198,702  \n",
      "4      chr7:127,685,612-127,687,611  \n",
      "     Gene ID              sequence                         Genomic_Location\n",
      "0       ACTB  CTTTCCTTCCCAGGGCGTGA  duplicate      chr7:5,528,404-5,530,403\n",
      "1       ACTB  CTTTCCTTCCCAGGGCGTGA        NaN      chr7:5,528,404-5,530,403\n",
      "2  HNRNPA2B1  AGCCACCAATAAAGAGCTTA        NaN    chr7:26,196,703-26,198,702\n",
      "3       SND1  CCCCTACGTAGGTCCTCTCA        NaN  chr7:127,685,612-127,687,611\n"
     ]
    }
   ],
   "source": [
    "# ## FULL 1 ALONE ##\n",
    " #USED FOR FINAL SAMPLE OUTPUT#\n",
    "\n",
    "get_genomic_location()\n",
    "\n",
    "guides_drop = guides.drop([0]).reset_index(drop=True) #drop SHAM row\n",
    "new_guides = guides_drop #rename\n",
    "\n",
    "#drop NA's and K gen loc returned from function 1, to generalize applicability\n",
    "guides_na = new_guides[new_guides.Genomic_Location == \"NA\"].reset_index(drop=True) #subset NA's\n",
    "guides_multiple = new_guides[new_guides.Genomic_Location.str.contains(\"K\")].reset_index(drop = True) #subset genomic loc matches, ie \"chr15_KI270851v1_alt:30,690-32,689\"\n",
    "guides_fail = [guides_na, guides_multiple] #merge guides that need to be done by hand\n",
    "pd.DataFrame(guides_fail).to_csv(\"Guides_To_Do_By_Hand\", sep = '\\t', index = False) #save failed guides as text file, \n",
    "                                                                                    #don't have good way to go straight to CSV maintaining structure\n",
    "\n",
    "guides_new_1 = new_guides[~new_guides.Genomic_Location.str.contains(\"K\")].reset_index(drop = True) #drop guides that must be done manually\n",
    "guides_new = guides_new_1[~guides_new_1.Genomic_Location.str.contains(\"NA\")].reset_index(drop = True) #drop any that didn't work\n",
    "print(guides_new)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## WORKING FULL 2 - with LOOP ##\n",
    "\n",
    "def get_primer_pairs():\n",
    "\n",
    "    from selenium import webdriver\n",
    "    from selenium.webdriver.common.keys import Keys\n",
    "    from webdriver_manager.chrome import ChromeDriverManager\n",
    "    from selenium.webdriver.support.ui import Select\n",
    "    from selenium.webdriver.common.by import By\n",
    "    from selenium.webdriver.support.ui import WebDriverWait\n",
    "    from selenium.webdriver.support import expected_conditions as EC\n",
    "    from webdriver_manager.chrome import ChromeDriverManager\n",
    "    import time\n",
    "    from selenium.common.exceptions import NoSuchElementException\n",
    "    \n",
    "    pair_1_seq = [] #empty lists for elements to go into\n",
    "    pair_2_seq = []\n",
    "    forward_Tm = []\n",
    "    reverse_Tm = []\n",
    "    product_length = []\n",
    "\n",
    "    for j in guides_new[\"Genomic_Location\"]: #feed in the genomic location from the dataframe through the loop\n",
    "\n",
    "        driver = webdriver.Chrome(ChromeDriverManager().install())\n",
    "        driver.get(\"https://www.ncbi.nlm.nih.gov/genome/gdv/browser/genome/?id=GCF_000001405.39\")\n",
    "        driver.minimize_window()\n",
    "        #assert \"NCBI's\" in driver.title\n",
    "        \n",
    "        time.sleep(3)\n",
    "\n",
    "        elem = driver.find_element_by_id(\"loc-search\") #find search bar\n",
    "        elem.clear() #clear search bar\n",
    "\n",
    "        elem.send_keys(j) #enter genomic location via j\n",
    "        elem.send_keys(Keys.RETURN)\n",
    "\n",
    "\n",
    "        wait = WebDriverWait(driver, 120) #create waitime to allow page to load\n",
    "\n",
    "        #primer blast via \"Tools\"\n",
    "        tools = wait.until(EC.element_to_be_clickable((By.XPATH, \"//a[@data-qtip='Tools']\"))) #select Tools dynamic id\n",
    "#         time.sleep(5) #slows down after ~25, if less no need for this pause\n",
    "        tools.click()\n",
    "\n",
    "        text1 = \"BLAST and Primer Search\";\n",
    "        drop1 = wait.until(EC.element_to_be_clickable((By.XPATH, \"//span[text()='\" +text1 +\"']\"))) #select option from dropdown\n",
    "        drop1.click()\n",
    "\n",
    "        text2 = \"Primer BLAST (Visible Range)\";\n",
    "        drop2 = wait.until(EC.element_to_be_clickable((By.XPATH, \"//span[text()='\" +text2 +\"']\"))) #select option from dropdown\n",
    "        drop2.click() #this will redirect to Primer BLAST page\n",
    "\n",
    "        driver.minimize_window()\n",
    "\n",
    "        #switch windows to new window\n",
    "        window_before = driver.window_handles[0] #starting window set\n",
    "        window_after = driver.window_handles[1] #new window set\n",
    "\n",
    "        driver.switch_to.window(window_after) #switch to new window to be able to intercat with it\n",
    "        print(driver.current_url) #check window switch successful\n",
    "        \n",
    "        time.sleep(10)\n",
    "\n",
    "        #set PCR product sizes\n",
    "        primemin = driver.find_element_by_name(\"PRIMER_PRODUCT_MIN\") #find search bar\n",
    "        primemin.clear() #clear search bar\n",
    "        primemin.send_keys(\"500\") #enter genomic location\n",
    "\n",
    "        primemax = driver.find_element_by_name(\"PRIMER_PRODUCT_MAX\") #find search bar\n",
    "        primemax.clear() #clear search bar\n",
    "        primemax.send_keys(\"1000\") #enter genomic location\n",
    "\n",
    "\n",
    "        ###find genomic positions to make 200bp around cut site###\n",
    "        forward_from = driver.find_element_by_id(\"PRIMER5_START\").get_attribute(\"value\") #value from gen browser\n",
    "\n",
    "        forward_to = driver.find_element_by_id(\"PRIMER5_END\") #will input new value here\n",
    "\n",
    "        forward_to_num = int(forward_from) + 900 #this adjusts to create 200bp buffer zone around CRISPR cut site\n",
    "        forward_to.send_keys(forward_to_num)\n",
    "\n",
    "        ##reverse##\n",
    "        reverse_to = driver.find_element_by_id(\"PRIMER3_END\").get_attribute(\"value\")\n",
    "\n",
    "        reverse_from = driver.find_element_by_id(\"PRIMER3_START\") #will input new value here\n",
    "\n",
    "        reverse_from_num = int(reverse_to) - 900 #this adjusts to create 200bp buffer zone around CRISPR cut site\n",
    "        reverse_from.send_keys(reverse_from_num)\n",
    "\n",
    "\n",
    "\n",
    "        #Select reference data base\n",
    "        database= Select(driver.find_element_by_id('PRIMER_SPECIFICITY_DATABASE'))\n",
    "        database.select_by_index(2) #\"Genomes for selected organisms (primary reference assembly only)\"\n",
    "\n",
    "        #Primer BLAST\n",
    "        get_primers = driver.find_element_by_id(\"b1\").click()\n",
    "        driver.minimize_window()\n",
    "        #time.sleep(5)\n",
    "        \n",
    "        time.sleep(120) # this sleep time is critical to allow website to load while finding primers\n",
    "       \n",
    "\n",
    "        pair_1 = driver.find_element_by_css_selector(\n",
    "        \"#alignments > div:nth-child(1) > table > tbody > tr:nth-child(2) > td:nth-child(2)\"\n",
    "        ).get_attribute(\"innerHTML\") #get text Forward Primer\n",
    "        #print(pair_1)\n",
    "            \n",
    "        pair_1_seq.append(pair_1) #add to empty list\n",
    "        \n",
    "        seq_1 = pd.DataFrame(pair_1_seq) #convert to data.Frame\n",
    "        guides_new[\"Pair 1\"] = seq_1 #add to csv\n",
    "\n",
    "\n",
    "        ##reverse##\n",
    "        pair_2 = driver.find_element_by_css_selector(\n",
    "        \"#alignments > div:nth-child(1) > table > tbody > tr:nth-child(3) > td:nth-child(2)\"\n",
    "        ).get_attribute(\"innerHTML\") #get text Reverse Primer\n",
    "        #print(pair_2)\n",
    "        pair_2_seq.append(pair_2)\n",
    "\n",
    "        seq_2 = pd.DataFrame(pair_2_seq)\n",
    "        guides_new[\"Pair 2\"] = seq_2\n",
    "\n",
    "        ##forward Tm##\n",
    "        pair_1_Tm = driver.find_element_by_css_selector(\n",
    "        \"#alignments > div:nth-child(1) > table > tbody > tr:nth-child(2) > td:nth-child(7)\"\n",
    "        ).get_attribute(\"innerHTML\") #get text Forward Tm\n",
    "\n",
    "        #print(pair_1_Tm)\n",
    "        forward_Tm.append(pair_1_Tm)\n",
    "\n",
    "        Tm_1 = pd.DataFrame(forward_Tm)\n",
    "\n",
    "        guides_new[\"Forward Tm\"] = Tm_1\n",
    "\n",
    "        ##reverse Tm##\n",
    "        pair_2_Tm = driver.find_element_by_css_selector(\n",
    "        \"#alignments > div:nth-child(1) > table > tbody > tr:nth-child(3) > td:nth-child(7)\"\n",
    "        ).get_attribute(\"innerHTML\") #get text Reverse Tm\n",
    "\n",
    "        #print(pair_2_Tm)\n",
    "        reverse_Tm.append(pair_2_Tm)\n",
    "\n",
    "        Tm_2 = pd.DataFrame(reverse_Tm)\n",
    "\n",
    "        guides_new[\"Reverse Tm\"] = Tm_2\n",
    "\n",
    "        ##product length##\n",
    "        product_len = driver.find_element_by_css_selector(\n",
    "        \"#alignments > div:nth-child(1) > table > tbody > tr:nth-child(5) > td\"\n",
    "        ).get_attribute(\"innerHTML\") #get text Product Length\n",
    "\n",
    "        #print(product_len)\n",
    "        product_length.append(product_len)\n",
    "\n",
    "        length = pd.DataFrame(product_length)\n",
    "\n",
    "        guides_new[\"Product Length\"] = length\n",
    "\n",
    "        driver.quit()\n",
    "        \n",
    "    #all of the output rows are shifted to account for empty output for first input (duplicate)   \n",
    "    guides_new['Pair 1'] = guides_new['Pair 1'].shift(+1)  \n",
    "    guides_new['Pair 2'] = guides_new['Pair 2'].shift(+1)\n",
    "    guides_new['Forward Tm'] = guides_new['Forward Tm'].shift(+1)\n",
    "    guides_new['Reverse Tm'] = guides_new['Reverse Tm'].shift(+1)\n",
    "    guides_new['Product Length'] = guides_new['Product Length'].shift(+1)\n",
    "    \n",
    "    print(guides_new)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Looking for [chromedriver 80.0.3987.106 mac64] driver in cache \n",
      "File found in cache by path [/Users/santiagoacero/.wdm/drivers/chromedriver/80.0.3987.106/mac64/chromedriver]\n",
      "https://www.ncbi.nlm.nih.gov/tools/primer-blast/index.cgi?ORGANISM=Homo%20sapiens&INPUT_SEQUENCE=NC_000007.14&PRIMER5_START=5528404&PRIMER3_END=5530403&PRIMER_PRODUCT_MAX=1999&SHOW_SVIEWER=true\n",
      "\n",
      "Looking for [chromedriver 80.0.3987.106 mac64] driver in cache \n",
      "File found in cache by path [/Users/santiagoacero/.wdm/drivers/chromedriver/80.0.3987.106/mac64/chromedriver]\n",
      "https://www.ncbi.nlm.nih.gov/tools/primer-blast/index.cgi?ORGANISM=Homo%20sapiens&INPUT_SEQUENCE=NC_000007.14&PRIMER5_START=5528404&PRIMER3_END=5530403&PRIMER_PRODUCT_MAX=1999&SHOW_SVIEWER=true\n",
      "\n",
      "Looking for [chromedriver 80.0.3987.106 mac64] driver in cache \n",
      "File found in cache by path [/Users/santiagoacero/.wdm/drivers/chromedriver/80.0.3987.106/mac64/chromedriver]\n",
      "https://www.ncbi.nlm.nih.gov/tools/primer-blast/index.cgi?ORGANISM=Homo%20sapiens&INPUT_SEQUENCE=NC_000007.14&PRIMER5_START=26196703&PRIMER3_END=26198702&PRIMER_PRODUCT_MAX=1999&SHOW_SVIEWER=true\n",
      "\n",
      "Looking for [chromedriver 80.0.3987.106 mac64] driver in cache \n",
      "File found in cache by path [/Users/santiagoacero/.wdm/drivers/chromedriver/80.0.3987.106/mac64/chromedriver]\n",
      "https://www.ncbi.nlm.nih.gov/tools/primer-blast/index.cgi?ORGANISM=Homo%20sapiens&INPUT_SEQUENCE=NC_000007.14&PRIMER5_START=127685612&PRIMER3_END=127687611&PRIMER_PRODUCT_MAX=1999&SHOW_SVIEWER=true\n",
      "     Gene ID              sequence                         Genomic_Location  \\\n",
      "0       ACTB  CTTTCCTTCCCAGGGCGTGA  duplicate      chr7:5,528,404-5,530,403   \n",
      "1       ACTB  CTTTCCTTCCCAGGGCGTGA        NaN      chr7:5,528,404-5,530,403   \n",
      "2  HNRNPA2B1  AGCCACCAATAAAGAGCTTA        NaN    chr7:26,196,703-26,198,702   \n",
      "3       SND1  CCCCTACGTAGGTCCTCTCA        NaN  chr7:127,685,612-127,687,611   \n",
      "\n",
      "                 Pair 1                Pair 2 Forward Tm Reverse Tm  \\\n",
      "0                   NaN                   NaN        NaN        NaN   \n",
      "1  TCATGGCCTTGTCACACGAG  CGGCTATTCTCGCAGCTCAC      60.32      61.21   \n",
      "2  TCATGGCCTTGTCACACGAG  CGGCTATTCTCGCAGCTCAC      60.32      61.21   \n",
      "3  CAGCACGTTTTGGCTCAACT  AGAATCACCACCAGAGGGGG      59.62      60.92   \n",
      "\n",
      "  Product Length  \n",
      "0            NaN  \n",
      "1            777  \n",
      "2            777  \n",
      "3            939  \n",
      "     Gene ID              sequence                   Genomic_Location  \\\n",
      "0       ACTB  CTTTCCTTCCCAGGGCGTGA  NaN      chr7:5,528,404-5,530,403   \n",
      "1  HNRNPA2B1  AGCCACCAATAAAGAGCTTA  NaN    chr7:26,196,703-26,198,702   \n",
      "2       SND1  CCCCTACGTAGGTCCTCTCA  NaN  chr7:127,685,612-127,687,611   \n",
      "\n",
      "                 Pair 1                Pair 2 Forward Tm Reverse Tm  \\\n",
      "0  TCATGGCCTTGTCACACGAG  CGGCTATTCTCGCAGCTCAC      60.32      61.21   \n",
      "1  TCATGGCCTTGTCACACGAG  CGGCTATTCTCGCAGCTCAC      60.32      61.21   \n",
      "2  CAGCACGTTTTGGCTCAACT  AGAATCACCACCAGAGGGGG      59.62      60.92   \n",
      "\n",
      "  Product Length  \n",
      "0            777  \n",
      "1            777  \n",
      "2            939  \n"
     ]
    }
   ],
   "source": [
    "## FULL 2 ALONE ##\n",
    "\n",
    "get_primer_pairs()\n",
    "\n",
    "guides_final = guides_new.drop([0]).reset_index(drop=True)\n",
    "print(guides_final)\n",
    "\n",
    "guides_final.to_csv(\"Guides_With_Genomic_Locations_and_Primers\", sep = '\\t', index = False) #save output into text file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## BOTH FUNCTIONS ##\n",
    "\n",
    "get_genomic_location()\n",
    "\n",
    "guides_drop = guides.drop([0]).reset_index(drop=True) #drop SHAM row\n",
    "new_guides = guides_drop #rename\n",
    "\n",
    "#drop NA's and K gen loc returned from function 1, to generalize applicability\n",
    "guides_na = new_guides[new_guides.Genomic_Location == \"NA\"].reset_index(drop=True) #subset NA's\n",
    "guides_multiple = new_guides[new_guides.Genomic_Location.str.contains(\"K\")].reset_index(drop = True) #subset genomic loc matches\n",
    "guides_fail = [guides_na, guides_multiple] #merge guides that need to be done by hand\n",
    "pd.DataFrame(guides_fail).to_csv(\"Guides_To_Do_By_Hand\", sep = '\\t', index = False) #save failed guides as text file\n",
    "\n",
    "guides_new_1 = new_guides[~new_guides.Genomic_Location.str.contains(\"K\")].reset_index(drop = True) #drop guides that must be done manually\n",
    "guides_new = guides_new_1[~guides_new_1.Genomic_Location.str.contains(\"NA\")].reset_index(drop = True) #drop any that didn't work\n",
    "print(guides_new)\n",
    "    \n",
    "get_primer_pairs()\n",
    "\n",
    "guides_final = guides_new.drop([0]).reset_index(drop=True)\n",
    "print(guides_final)\n",
    "\n",
    "\n",
    "guides_final.to_csv(\"Guides_With_Genomic_Locations_and_Primers_Final_2\", sep = '\\t', index = False) #save final product as text file"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
